src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -271,7 +282,28 @@ public void unexport(BodyReference ref) {
                 @SuppressFBWarnings(value="RV_RETURN_VALUE_IGNORED_BAD_PRACTICE", justification="runner.submit() result")
                 public Void call() throws Exception {
                     Jenkins j = Jenkins.getInstanceOrNull();
+                    if (j != null && !j.isQuietingDown() && execution != null && pausedByQuietMode.compareAndSet(true, false)) {
+                        try {
+                            execution.getOwner().getListener().getLogger().println("Resuming (Prepare for shutdown was canceled)");
+                        } catch (IOException e) {
+                            LOGGER.log(Level.WARNING, null, e);
+                        }
+                    }
                     if (paused.get() || j == null || (execution != null && j.isQuietingDown())) {
+                        if (j != null && j.isQuietingDown() && execution != null && pausedByQuietMode.compareAndSet(false, true)) {
+                            try {
+                                execution.getOwner().getListener().getLogger().println("Pausing (Preparing for shutdown)");
+                            } catch (IOException e) {
+                               LOGGER.log(Level.WARNING, null, e);
+                            }
+                            Timer.get().schedule(() -> {
+                                if (j.isQuietingDown()) {
+                                    Timer.get().schedule(this, Main.isUnitTest ? 1 : 10, TimeUnit.SECONDS);
```

Comment:
`this` is the `Callable` from line 281, I guess. That is what you wanted? If you meant to refer to the `Runnable` at line 299, you cannot use a lambda.

Reply:
Ha, I was wondering about that, I tried `Runnable.this` to disambiguate earlier just in case, but that was unsupported, and then forgot about it. I guess rerunning the `Callable` technically works, but yeah I'll just use an anonymous inner class instead.

Reply:
See d1322e2.